### PR TITLE
Level obstacles

### PR DESCRIPTION
**Pull request by Moneyl**
Changes:
- Level generation moved to `LevelGenerator.cs`. Makes a clear separation between level generation and level logic.
- Added time delta since last frame to `GameLogic.Update()`. Used for framerate independent code like player movement.
- All keys are now required to complete the end room and move into the next level. They're guaranteed to spawn somewhere in the level. Use the K key to add all keys to your inventory.
- Added special room types. Implemented by making classes that derive `Room` and implement custom per frame update functions. 
    - Windy rooms have constant wind that pushes the player around and periodically changes direction. 
    - Icy rooms have slippery floors and slower movement.
- Player movement is now based on velocity. This is so windy rooms can push the player around, and so players have to manage speed more carefully in icy rooms.
- Added floor spikes which damage the player when they collide. These only spawn in windy and icy rooms.
- Floor spikes are handled by a new class `LevelObject`. This is intended for interactive objects in the level that can't be picked up. Probably could merge this into `Item` and just have a "CanHold" flag.
- Fixed an issue with level cloning where the copy still referenced the original. I suspect this caused the bug where the wrong room would be rendered, then suddenly switch to the correct room. So far haven't been able to reproduce it again.

### CODE REVIEW

File: src/logic/GameLogic.cs [hunk 1]
```diff
@@ -60,15 +61,39 @@ public class GameLogic {
...
+			//Player died, reset level
+			if (Player.Health <= 0) {
+				Level.CurrentRoom.OnExit(Level, Level.StartRoom); //Call on exit so things like ambient music are stopped
+				Sounds.PlaySound("assets/sounds/PlayerDeath0.wav"); //Play death sound
+				//Recreate player and generate new level
+				Player = new Player(new Vector2(0.0f, 0.0f));
+				Level = LevelGenerator.TryGenerateLevel();
+				Level.Player = Player;
+				Console.WriteLine($"Player died with a score of {Level.Score}!");
+				Renderer.EventQueue.Enqueue("LevelRegenerated"); //Signal to renderer to regenerate map scene
```
Comment (trentv4) [inline src/logic/GameLogic.cs:92]: Later, this should be changed to a "PlayerDeath" so can I can implement a death screen. For now, good.

File: src/tests/LevelTests.cs [hunk 2]
```diff
@@ -8,11 +8,11 @@ namespace Project.Tests {
...
-            //Ensure that all rooms have at least 2 connections so the player can't get stuck
-            foreach (var room in level.Rooms) {
+			//Generate a new level
+			Level level = LevelGenerator.TryGenerateLevel();
+
+			//Ensure that all rooms have at least 2 connections so the player can't get stuck
+			foreach (var room in level.Rooms) {
                 Assert.IsTrue(room.ConnectedRooms.Length >= 2);
             }
         }
```
Comment (trentv4) [inline src/tests/LevelTests.cs:18]: I would add a test where it generates the level about 1000+ times to get a little extra security.

File: src/logic/Level.cs [hunk 3]
```diff
@@ -493,21 +238,155 @@ public enum VisitedState {
...
+			base.OnExit(level, nextRoom);
+			level.Player.FloorFriction = Player.DefaultFloorFriction;
+			level.Player.MovementSpeed = Player.DefaultMovementSpeed;
+		}
+	}
+
+	/// <summary> Interactable object found in rooms. </summary>
+	public class LevelObject : ICloneable {
+		public Vector2 Position;
+		public readonly string TextureName;
```
Comment (trentv4) [inline src/logic/Level.cs:340]: This should be a reference to some sort of model loading after the model loading pull request is merged.

**Review by trentv4 — CHANGES_REQUESTED** [review]

File: src/logic/Player.cs [hunk 4]
```diff
@@ -12,6 +12,16 @@ public class Player : ICloneable {
 		public int Armor = 0;
 		public int CarryWeight = 10;
 		public Vector2 Position;
+		public Vector2 Velocity;
+		public float MovementSpeed = DefaultMovementSpeed;
+		public readonly float MaxSpeed = 6.5f;
+		public float FloorFriction = DefaultFloorFriction;
```
[the remark this answers is not in the document]
Comment (Moneyl) [reply src/logic/Player.cs]: Agreed, doesn't make sense to have that in `Player`. Moved to `Room` and it now passes the current room to `Player.Update()` so they can access it.

File: src/tests/LevelTests.cs (hunk 2 above)
Comment (Moneyl) [reply src/tests/LevelTests.cs:18]: Done.

File: src/util/ListExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        /// <summary>
+        /// Override for List<T>.Contains() that uses a predicate instead of comparing each item with a value.
+        /// Useful when checking if a list contains an object with specific member values.
+        /// </summary>
+        public static bool Contains<T>(this List<T> list, Predicate<T> match) {
+            //List<T>.Find() returns null if the predicate doesn't match any value
+            return list.Find(match) != null;
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (Moneyl) [reply src/util/ListExtensions.cs]: Done.

**Pull request merged**